torch/export/dynamic_shapes.py

```diff
@@ -904,6 +909,149 @@ def check_shape(path, t, dynamic_shape):
+    process of producing constraints, and now both use assume_static_by_default=False.
+
+    For _dynamo.export(), the semantics for dynamic_shapes are:
+    - None: dynamic, allocated a symbol
+    - Dim/DerivedDim: a strict assertion on the min/max range for this symbol, and require a specification
+      for all dims governed by this symbol (i.e. relations, equality, linear relations, etc.)
+
+    For export.export(), historically dynamism for unspecified dims has been undesirable, so the semantics are:
+    - True: dynamic, allocated a symbol
+    - None/unspecified: static
+    - Dim/DerivedDims: also a strict assertion
+
+    To allow both APIs to follow the same process for producing constraints, this function converts dynamic_shapes
+    for export.export() to be compatible with _process_dynamic_shapes() and assume_static_by_default=False, turning them
+    into essentially what they'd look like for _dynamo.export().
+
+    An example conversion might look like, for a 3-d input tensor:
+
+        input spec: {
+            0: True,
+            1: None,
+            2: Dim("dx"),
+        }
+        output spec: {
+            0: None,  # None: dynamic by default
+            1: 32,  # explicitly provide static shape
+            2: Dim("dx"),  # remains the same
+        }
+    """
+
+    def _tree_map_helper(tree, val):
+        """
+        If the user generally specifies dynamic_shapes=None for a pytree input,
+        we'd like to convert this into a tree of Nones following the input spec,
+        so we can explicitly specify static dims for all tensor dimensions.
+        Non-builtin types for pytree (e.g. custom dataclasses) creates some difficulty,
+        in which case the correct format is a list containing specs for each child attribute.
+        """
+        if (node_type := _get_node_type(tree)) not in SUPPORTED_NODES:  # is_leaf
+            return val
+        flatten_fn = SUPPORTED_NODES[node_type].flatten_fn
+        child_pytrees, context = flatten_fn(tree)  # flatten from whatever original type
+        unflatten_fn = SUPPORTED_NODES[
+            node_type if node_type in BUILTIN_TYPES else list
+        ].unflatten_fn
+        children = [_tree_map_helper(child, val) for child in child_pytrees]
+        return unflatten_fn(
+            children, context
+        )  # unflatten into original type, or list if not built-in type
+
+    if dynamic_shapes is True:  # dynamic by default
+        return None
+    elif (
+        dynamic_shapes is None or len(dynamic_shapes) == 0
+    ):  # create pytree structure of static dim
+        dynamic_shapes = _tree_map_helper(combined_args, None)
+    if isinstance(dynamic_shapes, (tuple, list)):
+        combined_args = type(dynamic_shapes)(combined_args.values())  # type: ignore[assignment, misc]
+
+    def transform_shapes(path, tensor, shape):
+        def _marked_dynamic(tensor, i):
```

Comment:
Oh boy, we have to handle these as well in export? I'd really prefer not to. If you have to, can you describe the semantics here (also add comment)?

Reply:
Yeah, it's used in some AOTInductor test cases, you could leave `dynamic_shapes` as None, but call `mark_dynamic(tensor, index=i)`

Reply:
Hmm sorry might be a dumb question: couldn't we rewrite AOTInductor cases to use the proper dynamic shapes API? I am afraid we are making this logic too complicated hahaa

Reply:
I agree. I don't think we want to be in the business of defining what `mark_dynamic` does in export: it looks like it has the same effect as `True`? So in principle it should be possible to update those tests. If that's too much scope for this PR, we should definitely track a follow up.

Reply:
There was a small moment I was afraid all the changes in this PR would be unnecessary, and we only needed to switch to calling `mark_dynamic` under the hood to implement auto dynamic shapes, but seems like currently it is different from `True`, and slightly different from creating a plain `Dim()`:

- Unlike True, it doesn't automatically specialize or constrain itself, so you'll get a constraint violation
- Unlike Dim, if you mark multiple dims dynamic and they turn out to be equal, they'll get the symbol without complaining

There seems to be a lot of tests that rely on this (AOTInductor, torchbench, probably some internal), so I'll try to maintain support as is for this PR, and then we can decide to deprecate after?

Reply:
Sounds good.

Reply:
Could you leave a TODO in the code and file a task?